✨ Initial support for Context Menus

NinjaLabs-Dev opened the pull request:
This aims to add Message and User context menu support to slash commands, I'm unsure how we'd want to structure this, for example having a separate folder to contain these types of interactions or build onto #71 and use a parent Interactions folder to cover all bases?

Current Issues:
While this system with the existing interaction handlers, the issue I have met is regarding registration, specifically within the `SlashCommand` create() function, it appears the discord-php command builder doesn't allow creation of commands without a description. I did have a look at the source code of the library and other relating issues like [this one](https://github.com/discord-php/DiscordPHP/issues/691) however I'm personally unsure if this would require a pull request to adjust the command builder itself. 

I did attempt to mitigate this with the follow code:

```php
    $command = CommandBuilder::new()
        ->setName($this->getName())
        ->setType($this->getType());
        
    if($this->getType() !== DiscordCommand::CHAT_INPUT) {
        $command = $command->setDescription($this->getDescription());
    }
```

However, this creates a typed property error as $description is accessed before initialization (This relates to the `CommandBuilder` within discord-php)

I may have missed the mark somewhere along the line of attempting to add this support, however feedback is appreciated.

Log1x commented:
nice! i think using the bot description as a fallback if the type isn't a `CHAT_INPUT` should be ok. what if we allow an array of types so you can do a normal slash command as well as a context menu from the same command? if that seems sane and we can loop/register those, i'd say keep it all together perhaps as far as folder goes.

NinjaLabs-Dev commented:
My mistake on the description fallback, I intended for this to add the description to the command if it _was_ a `CHAT_INPUT`, an array would be a great idea, I actually did on my first attempt of creating my bot with Node.js however opted for the convenience of a Laravel environment.

---

While typing this comment, I've had an idea how we can get around the access / builder issue regarding the context menus.

The builder is the class causing this issue as it requires a description, though Context Menus are not allowed to have descriptions (likely as this wasn't the indented use).

Instead, to navigate the issue and not require a pull request to discord-php (which could still be done but not required) we can unset the description after we've setup all our requirements. See the below example

```php
$command = CommandBuilder::new()
    ->setName($this->getName())
    ->setType($this->getType())
    ->setDescription($this->getDescription());

if ($permissions = $this->getPermissions()) {
    $command = $command->setDefaultMemberPermissions($permissions);
}

if ($this->getRegisteredOptions()) {
    foreach ($this->getRegisteredOptions() as $option) {
        $command = $command->addOption($option);
    }
}

$command = $command->toArray();
if($this->getType() !== DiscordCommand::CHAT_INPUT) {
    unset($command['description']);
}

$command = collect($commandArray)
    ->put('guild_id', $this->getGuild())
    ->filter()
    ->all();

return new DiscordCommand($this->discord(), $command);
```

Honestly, it's not the cleanest solution, as that would be to update the builder however it reaches the goal. (I have tested this also with a positive result)

Log1x commented:
> Instead, to navigate the issue and not require a pull request to discord-php (which could still be done but not required) we can unset the description after we've setup all our requirements. See the below example

Ahh I misunderstood the original issue. That makes sense & seems fine for now. Should be an easy PR into Discord PHP and removal of that in Laracord in the future. Feel free to implement that stuff & let me know when you want me to come in and finish up! Appreciate it.

NinjaLabs-Dev commented:
I've got this sorted to a point.

Currently, the create function returns a DiscordCommand class, how do you suggest we work around this return if the command has multiple types, I haven't looked enough into this framework / laracord to fully understand the life cycle and how it integrates into the handle function.

Another thought also, with the context interaction not having options unlike chat text options, how do you suggest we alter the `->value` helper to allow developers to easily access the targeted message? Currently, I've just set the parsedOptions array to empty if it's not type `CHAT_INPUT` as this causes an error, mainly as `$interaction->data->options` doesn't exist on context interactions (for obvious reason).

Log1x commented:
Hmm. Looking into this more and with how limiting context menus are compared to actual slash commands – I'm wondering if it is actually worth the headache to allow the array of types or if we should in fact just do a separate, more appropriate parent class/folder for `ContextMenus`.

NinjaLabs-Dev commented:
As you can see from the above commit I've done more work into adding this support, I've created the ContextMenu command class along with make commands and stubs, currently this uses the existing createSlashCommands function however has checks for the type of command, the main reason I decided to do this was due to them basically being the same structure and code for the ContextMenus with some exceptions with options.

I've also added a `getCleanName` function to the ContextMenu class as Discord-PHP's command validation doesn't support the format within names of ContextMenus (normal `CHAT_INPUT`s don't support spaces or mixed cases whereas the other two types do)

I'm unsure if we'd want to completely separate the two creations however I'm unsure how effective that will be.

Log1x commented:
nice! will jump in on this tonight/tomorrow.

NinjaLabs-Dev commented:
Sounds good, I think one thing left to consider is how to access the target values from the interaction. As we've spoken about ContextMenus don't support options however do return the target, for example for a `MESSAGE` type it returns the message and subsequently the author, guild and other data.

NinjaLabs-Dev commented:
I did some more testing around, while trying to protect the correct label within Discord this also introduced a little issue while trying to reference back to the interactions and the general command response, I've updated the flow of how it registers the interaction to Discord to stop this issue while also keeping the correct display name. 

I think once Discord-PHP supports context menus natively, this likely will become a novelty and won't be needed.

I've also removed some of my testing code, prints and generally removed functions not needed for the context menus.

NinjaLabs-Dev commented:
Great to hear, only thing I can notice from the changes is the Context Menu stub registers $type as a string however needs to be a string|int.

Log1x commented:
fixed, good catch!

NinjaLabs-Dev commented:
I also seem to get a `Call to undefined method App\Bot::getInteractions()` error attempting to run this too, unsure if you've forgotten to push this or if it's intended for another function. 

It's within the `handleInteractions()` function in Laracord:732

Log1x commented:
kk will fix it with the cache stuff momentarily.
